nncf/experimental/torch/nas/bootstrapNAS/elasticity/base_handler.py

```diff
@@ -0,0 +1,248 @@
+
+        :return: elasticity configuration
+        """
+
+    @abstractmethod
+    def get_maximum_config(self) -> ElasticityConfig:
+        """
+        Forms an elasticity configuration that describes a Subnet with maximum elastic values
+
+        :return: elasticity configuration
+        """
+
+    @abstractmethod
+    def activate_supernet(self) -> None:
+        """
+        Activates the Supernet - the original network to which elasticity was applied.
+        The Supernet is usually a maximum Subnet. But it might be not the case if maximum value of elastic properties
+        is limited for efficiency of search. For instance, a supernet may contain 32 convolutional channels,
+        but maximum subnet can be restricted to have 24 channels, at maximum.
+        """
+
+    @abstractmethod
+    def set_config(self, config: ElasticityConfig) -> None:
+        """
+        Activates a Subnet that corresponds to the given elasticity configuration
+
+        :param config: elasticity configuration
+        """
+
+    @abstractmethod
+    def get_state(self) -> Dict[str, Any]:
+        """
+        Returns a dictionary with Python data structures (dict, list, tuple, str, int, float, True, False, None) that
+        represents state of the object.
+
+        :return: state of the object
+        """
+
+    @abstractmethod
+    def load_state(self, state: Dict[str, Any]) -> None:
+        """
+        Initializes object from the state.
+
+        :param state: Output of `get_state()` method.
+        """
+
+    def activate_random_subnet(self) -> None:
+        """
+        Activates a Subnet with random values of elastic properties.
+        """
+        config = self.get_random_config()
+        self.set_config(config)
+
+    def activate_minimum_subnet(self) -> None:
+        """
+        Activates a minimum Subnet that corresponds to the minimum values of elastic properties.
+        """
+        config = self.get_minimum_config()
+        self.set_config(config)
+
+    def activate_maximum_subnet(self) -> None:
+        """
+        Activates a maximum Subnet that corresponds to the maximum values of elastic properties.
+        """
+        config = self.get_maximum_config()
+        self.set_config(config)
+
+
+class SEHandlerStateNames:
+    ACTIVE_CONFIG = 'active_config'
+
+
+class SingleElasticityHandler(ElasticityHandler, ABC):
+    """
+    An interface for handling a single elasticity dimension in the network, e.g. elastic width or depth.
+    """
+    _state_names = SEHandlerStateNames
+
+    @abstractmethod
+    def get_search_space(self) -> ElasticSearchSpace:
+        """
+        :return: search space that is produced by iterating over all elastic parameters
+        """
+
+    @abstractmethod
+    def get_kwargs_for_flops_counting(self) -> Dict[str, Any]:
+        """
+        Provides arguments for counting flops of the currently activated subnet.
```

Comment:
Honestly, It's not a clear interface.

Reply:
I can introduce `FlopsCountingArgs` class and refactor this method to fill this struct:
```python
def get_kwargs_for_flops_counting(self, args: FlopsCountingParams) -> FlopsCountingParams:
    args.var1=value
    return args
```
@alexsu52 Would it be more clear?

Reply:
It's more clear. As I understand you need a list of layers with attributes to calculate Flops. What do you think about returning some subnet descriptor which includes layers and their parameters?

Reply:
I don't feel comfortable to open internals of Elastic Handlers now. I intentionally encapsulated all elasticity related operations and structures inside them. They don't look simple and easy to use outside, moreover incorrect usage may potentially lead to undefined behavior.
```python
node_name_vs_dynamic_input_width_op_map: Dict[NNCFNodeName, ElasticWidthOp],
pruned_module_groups_info: Clusterization[ElasticWidthInfo](id_fn=lambda x: x.node_name),
...
skipped_blocks: BuildingBlocks,
skip_dependencies: GroupedBlockIDs,
node_names_per_block: Dict[int, List[NNCFNodeName]],
ordinal_ids: OrdinalIDs
```

Instead, I propose to introduce a single simple method for each handler:
```python
kernel_sizes = self.kernel_handler.get_active_kernel_sizes_per_node()
names_of_skipped_nodes = self.depth_handler.get_names_of_skipped_nodes()
input_width_values, output_width_values = self.width_handler.get_active_in_out_width_values()
```
If more information is needed outside of handler, it will be not difficult to provide.